File: fly-http/src/main/java/com/xia/fly/http/callback/Callback.java
package com.xia.fly.http.callback;

import android.content.Context;

import java.lang.ref.WeakReference;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LifecycleOwner;
import io.reactivex.disposables.Disposable;
import okhttp3.ResponseBody;

/**
 * @author xia
 * @date 2018/7/27.
 */
@SuppressWarnings("WeakerAccess")
public abstract class Callback<T> {
    private WeakReference<Context> mContext;
    public final LifecycleOwner mLifecycleOwner;

    public Context getContext() {
        return mContext != null ? mContext.get() : null;
    }

    public Callback(@NonNull LifecycleOwner lifecycleOwner) {
        mLifecycleOwner = lifecycleOwner;
        Context context = null;
        if (lifecycleOwner instanceof AppCompatActivity) {
            context = ((AppCompatActivity) lifecycleOwner);
        } else if (lifecycleOwner instanceof Fragment) {
            context = ((Fragment) lifecycleOwner).getContext();
        }
        mContext = context != null ? new WeakReference<>(context) : null;
    }

    public void onSubscribe(Disposable d) {
    }

    public void onError(Throwable e) {
    }

    public void onComplete() {
    }

    public void inProgress(float progress, long total) {
    }

    public abstract void parseNetworkResponse(ResponseBody responseBody) throws Exception;

    public abstract void onResponse(T response);
}

File: fly-http/src/main/java/com/xia/fly/http/callback/Callback.kt
package com.xia.fly.http.callback

import android.content.Context
import androidx.lifecycle.LifecycleOwner
import com.xia.fly.utils.FlyUtils
import io.reactivex.disposables.Disposable
import okhttp3.ResponseBody
import java.lang.ref.WeakReference

/**
 * @author xia
 * @date 2018/7/27.
 */
@Suppress("UNUSED_PARAMETER")
abstract class Callback<T>(private val lifecycleOwner: LifecycleOwner) {
    private var mContext: WeakReference<Context>? = null

    init {
        val context: Context? = FlyUtils.getContext(lifecycleOwner)
        context?.let {
            mContext = WeakReference(it)
        }
    }

    fun getContext(): Context? {
        return mContext?.get()
    }

    fun getLifecycleOwner(): LifecycleOwner {
        return lifecycleOwner
    }

    fun onSubscribe(d: Disposable) {}

    fun onError(e: Throwable) {}

    fun onComplete() {}

    fun inProgress(progress: Float, total: Long) {}

    @Throws(Exception::class)
    abstract fun parseNetworkResponse(responseBody: ResponseBody)

    abstract fun onResponse(response: T?)
}
